docs/src/literate/topology_optimization.jl

```diff
@@ -0,0 +1,502 @@
+# will later apply a force in negative y-direction.
+
+function create_grid(n) 
+    corners = [Vec{2}((0.0, 0.0)),
+               Vec{2}((2.0, 0.0)),
+               Vec{2}((2.0, 1.0)),
+               Vec{2}((0.0, 1.0))]
+    grid = generate_grid(Quadrilateral, (2*n, n), corners);
+    
+    ## node-/facesets for boundary conditions
+    addnodeset!(grid, "clamped", x -> x[1] ≈ 0.0)
+    addfaceset!(grid, "traction", x -> x[1] ≈ 2.0 && norm(x[2]-0.5) <= 0.05); 
+    return grid
+end
+#md nothing # hide
+
+# Next, we create the FE values, the DofHandler and the Dirichlet boundary condition.
+
+function create_values()
+    ## quadrature rules
+    qr      = QuadratureRule{2,RefCube}(2)
+    face_qr = QuadratureRule{1,RefCube}(2)
+
+    ## geometric interpolation
+    interpolation_geom = Lagrange{2,RefCube,1}()
+
+    ## cell and facevalues for u
+    cellvalues = CellVectorValues(qr, Lagrange{2,RefCube,1}(), Lagrange{2,RefCube,1}())
+    facevalues = FaceVectorValues(face_qr, Lagrange{2,RefCube,1}(), Lagrange{2,RefCube,1}())
+    
+    return cellvalues, facevalues
+end
+
+function create_dofhandler(grid)
+    dh = DofHandler(grid)
+    push!(dh, :u, 2, Lagrange{2,RefCube,1}()) ## displacement
+    close!(dh)
+    return dh
+end
+
+function create_bc(dh) 
+    dbc = ConstraintHandler(dh)
+    add!(dbc, Dirichlet(:u, getnodeset(dh.grid, "clamped"), (x,t) -> zero(Vec{2}), [1,2]))
+    close!(dbc)
+    t = 0.0
+    update!(dbc, t)
+    return dbc
+end
+#md nothing # hide
+
+# Now, we define a struct to store all necessary material parameters (stiffnes tensor of the bulk material
+# and the parameters for topology optimization) and add a constructor to the struct to
+# initialize it by using the common material parameters Young's modulus and Poisson number.
+
+struct MaterialParameters{T, S <: SymmetricTensor{4, 2, T}}
+    C::S
+    χ_min::T 
+    p::T
+    β::T
+    η::T
+end
+#md nothing # hide
+
+function MaterialParameters(E, ν, χ_min, p, β, η) 
+    δ(i,j) = i == j ? 1.0 : 0.0 ## helper function
+
+    G = E / 2(1 + ν) ## =μ
+    λ = E*ν/(1-ν^2) ## correction for plane stress included
+
+    temp(i,j,k,l) = λ * δ(i,j)*δ(k,l) + G* (δ(i,k)*δ(j,l) + δ(i,l)*δ(j,k)) 
+    
+    C = SymmetricTensor{4, 2}(temp)
+    return MaterialParameters(C, χ_min, p, β, η)
+end
+#md nothing # hide
+
+# To store the density and the strain required to calculate the driving forces, we create the struct
+# `MaterialState`. We add a constructor to initialize the struct. The function `UpdateMaterialStates!`
+# updates the density values once we calculated the new values.
+
+mutable struct MaterialState{T, S <: AbstractArray{SymmetricTensor{2, 2, T}, 1}} 
+    χ::T ## density
+    ε::S ## strain in each quadrature point
+end
+
+function MaterialState(ρ, n_qp)
+    return MaterialState(ρ, Array{SymmetricTensor{2,2,Float64},1}(undef, n_qp))
+end
+
+function UpdateMaterialStates!(χn1, states, dh)
```

Comment:
This is capitalized like a constructor, but it is a regular function.
```suggestion
function update_material_states!(χn1, states, dh)
```